spinoso-string/src/lib.rs

```diff
@@ -1440,149 +1276,47 @@ impl String {
-                let mut bytes = self.buf.as_slice();
-
-                while !bytes.is_empty() {
-                    let (ch, size) = bstr::decode_utf8(bytes);
-                    if let Some(ch) = ch {
-                        // Converting a UTF-8 character to lowercase may yield
-                        // multiple codepoints.
-                        for ch in ch.to_lowercase() {
-                            replacement.push_char(ch);
-                        }
-                        bytes = &bytes[size..];
-                    } else {
-                        let (substring, remainder) = bytes.split_at(size);
-                        replacement.extend_from_slice(substring);
-                        bytes = remainder;
-                    }
-                }
-                self.buf = replacement;
-            }
-        }
+        self.inner.make_lowercase();
     }
 
     /// Modify this `String` to have the all characters converted to uppercase.
     #[inline]
     pub fn make_uppercase(&mut self) {
-        match self.encoding {
-            Encoding::Ascii | Encoding::Binary => {
-                self.buf.make_ascii_uppercase();
-            }
-            Encoding::Utf8 => {
-                // This allocation assumes that in the common case, upper-casing
-                // `char`s do not change the length of the `String`.
-                let mut replacement = Vec::with_capacity(self.buf.len());
-                let mut bytes = self.buf.as_slice();
-
-                while !bytes.is_empty() {
-                    let (ch, size) = bstr::decode_utf8(bytes);
-                    if let Some(ch) = ch {
-                        // Converting a UTF-8 character to lowercase may yield
-                        // multiple codepoints.
-                        for ch in ch.to_uppercase() {
-                            replacement.push_char(ch);
-                        }
-                        bytes = &bytes[size..];
-                    } else {
-                        let (substring, remainder) = bytes.split_at(size);
-                        replacement.extend_from_slice(substring);
-                        bytes = remainder;
-                    }
-                }
-                self.buf = replacement;
-            }
-        }
+        self.inner.make_uppercase();
     }
 
     #[inline]
     #[must_use]
     #[cfg(feature = "casecmp")]
     #[cfg_attr(feature = "docsrs", doc(cfg(feature = "casecmp")))]
     pub fn ascii_casecmp(&self, other: &[u8]) -> Ordering {
-        focaccia::ascii_casecmp(self.buf.as_slice(), other)
+        focaccia::ascii_casecmp(self.as_slice(), other)
     }
 
     #[inline]
     #[must_use]
     #[cfg(feature = "casecmp")]
     #[cfg_attr(feature = "docsrs", doc(cfg(feature = "casecmp")))]
     pub fn unicode_casecmp(&self, other: &String, options: CaseFold) -> Option<bool> {
-        let left = self.buf.as_slice();
+        let left = self;
         let right = other;
```

Comment:
I think there's some deref and type inference magic happening here. Can we change these lines to the following to make it a bit more clear what is happening?

```suggestion
        let left = self.as_slice();
        let right = other.as_slice();
```